provider.go

```diff
@@ -5,110 +5,189 @@ package azuresecrets
-	"github.com/Azure/go-autorest/autorest"
-	"github.com/Azure/go-autorest/autorest/azure/auth"
-	"github.com/hashicorp/vault-plugin-secrets-azure/api"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/vault/sdk/logical"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization"
+
 	"github.com/hashicorp/vault/sdk/helper/useragent"
+	"github.com/microsoftgraph/msgraph-sdk-go/models"
+
+	"github.com/hashicorp/vault-plugin-secrets-azure/api"
 )
 
-var _ api.AzureProvider = (*provider)(nil)
+// AzureProvider is an interface to access underlying Azure Client objects and supporting services.
+// Where practical the original function signature is preserved. Client provides higher
+// level operations atop AzureProvider.
+type AzureProvider interface {
+	api.ApplicationsClient
+	api.GroupsClient
+	api.ServicePrincipalClient
+
+	CreateRoleAssignment(
+		ctx context.Context,
+		scope string,
+		roleAssignmentName string,
+		parameters armauthorization.RoleAssignmentCreateParameters) (armauthorization.RoleAssignmentsClientCreateResponse, error)
+	DeleteRoleAssignmentByID(ctx context.Context, roleID string) (armauthorization.RoleAssignmentsClientDeleteByIDResponse, error)
+	ListRoleAssignments(ctx context.Context, filter string) ([]*armauthorization.RoleAssignment, error)
+	ListRoleDefinitions(ctx context.Context, scope string, filter string) (result []*armauthorization.RoleDefinition, err error)
+	GetRoleDefinitionByID(ctx context.Context, roleID string) (result armauthorization.RoleDefinitionsClientGetByIDResponse, err error)
+}
+
+var _ AzureProvider = (*provider)(nil)
 
 // provider is a concrete implementation of AzureProvider. In most cases it is a simple passthrough
 // to the appropriate client object. But if the response requires processing that is more practical
 // at this layer, the response signature may different from the Azure signature.
 type provider struct {
 	settings *clientSettings
+	// HTTP Client used to make role assignment and definition requests
+	// using new arm libraries
+	httpClient *http.Client
 
 	appClient    api.ApplicationsClient
 	spClient     api.ServicePrincipalClient
 	groupsClient api.GroupsClient
-	raClient     *authorization.RoleAssignmentsClient
-	rdClient     *authorization.RoleDefinitionsClient
+	raClient     *armauthorization.RoleAssignmentsClient
+	rdClient     *armauthorization.RoleDefinitionsClient
 }
 
 // newAzureProvider creates an azureProvider, backed by Azure client objects for underlying services.
-func newAzureProvider(settings *clientSettings, passwords api.Passwords) (api.AzureProvider, error) {
+func newAzureProvider(settings *clientSettings, passwords api.Passwords) (AzureProvider, error) {
```

Comment:
This is unrelated to your PR (sorry!) but probably a good time to figure out what's going on. It appears that `passwords` is unused here. Looking closer, it appears that our password `Generate()` func in [api/passwords.go](https://github.com/hashicorp/vault-plugin-secrets-azure/blob/VAULT-11806/msgraph-sdk-upgrade/api/passwords.go#L26) is also unused. It doesn't looks like Azure lets you provide a password but instead returns one to the client. This leads me to wonder if/how [password_policy](https://github.com/hashicorp/vault-plugin-secrets-azure/blob/main/path_config.go#L76) works? 🤔

Reply:
It doesn't appear as though it does work. I wonder if we need to use [SetSecretText](https://github.com/microsoftgraph/msgraph-sdk-go/blob/9f7263506f34618b4206543ff2115e96b1233d72/models/password_credential.go#L328) in AddApplicationPassword()?

Reply:
Based on their API docs:
1. https://learn.microsoft.com/en-us/graph/api/application-addpassword?view=graph-rest-1.0&tabs=http
2. https://learn.microsoft.com/en-us/graph/api/serviceprincipal-addpassword?view=graph-rest-1.0&tabs=http

It doesn't seem like there's a way to set the password.

Maybe the fields were just added in case their API adds support for a user defined password in the future.

Reply:
Hmm, ya it doesn't look like you can set a password even with that `SetSecretText()` method. I think it's safe to say password policies don't work. I'm thinking we'll want to deprecate the parameter, remove the code, and announce this in documentation / upgrade guides at minimum. Can be done in a separate task if that's preferred.

Reply:
Thanks for the context folks! Will add a ticket to track this in our backlog, and going to opt to tackle this in a follow-up PR 👍🏼